### PR TITLE
[FLINK-7461] Remove Backwards compatibility with <= Flink 1.1

### PR DESCRIPTION
**Pull request by StefanRRichter**
## Brief change log

This PR removes backwards compatibility with Flink <= 1.1 from the code. In particular, we can remove many special cases, such as the non-partitionable state that was created by the `Checkpointed` interface (replaced by `CheckpointedFunction` in 1.2).

We also drop aligned windows, which have already been a "hidden feature" since 1.2, because they still rely on the outdated `Checkpointed` interface.

The `Checkpointed` and `CheckpointedRestoring` interfaces are also removed and all related test cases have either been deleted or adapted.

Edit: as @kl0u pointed out, this also breaks direct compatibility with Flink 1.2 for the CEP library. Users can still upgrade through Flink 1.2 -> 1.3 -> 1.4.

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes, SavepointSerializer)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha @StephanEwen ; the relevant commit is only 95e44099784c9deaf2ca422b8dfc11c3d67d7f82 .

**Comment by kl0u** [comment]
From a brief check, I see that this also removed backwards compatibility also for 1.2 for the CEP library. The reason is that the CEP library in Flink 1.2 was the same as in Flink 1.1 (no upgrade) and so it was using the old state abstractions.

Given that in the ML we discussed about dropping compatibility with 1.1 and agreed on that, a library should not affect that development, but there must be an explicit discussion in the ML for this specific matter before merging it. Even if this just means just saying that you have to go through 1.3 for CEP if you want to migrate from 1.2 to 1.4.

**Comment by StefanRRichter** [comment]
Yes, that is true, but I think it should because otherwise there is still no way to remove the `Checkpointed` / legacy state related code paths.

**Comment by kl0u** [comment]
I agree that this is a change that we agreed upon. My only point is that we have to communicate it also in the ML. Probably with a thread that points to this JIRA and the PR and mentions this issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java [hunk 1]
```diff
@@ -49,27 +49,8 @@ public static long getStateSize(StateObject handle) {
...
-					}
-				}
-			}
-
-			if (exception != null) {
-				throw exception;
-			}
-		}
+		Iterable<? extends StateObject> handlesToDiscard) throws Exception {
+		LambdaUtil.applyToAllWhileSuppressingExceptions(handlesToDiscard, StateObject::discardState);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java]: Is this change really necessary?

**Comment by dawidwys** [comment]
Just wanted to add to what @kl0u said that if we remove the compatibility with <= 1.2 of CEP library  there will be lot more code that will be no longer used than just that in this PR. E.g.

- `NFA#migrateNFA`
- `NFACompiler#migrateGraph`
- `SharedBuffer#migrateSharedBuffer`
- `SharedBuffer#readObject`

just to name a few. I think it would be really important to create a following JIRA to introduce those changes in CEP, if we proceed with this PR.

**Comment by StefanRRichter** [comment]
@dawidwys It is of course possible that I missed some dead code between. I can also just remove the parts you suggested for this PR.

**Comment by dawidwys** [comment]
@StefanRRichter I am ok either with you removing it in this PR or with creating a separate PR, that e.g. I could work on. Just wanted to make sure we will not miss it.

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/testutils/ExactlyOnceValidatingConsumerThread.java [hunk 2]
```diff
@@ -95,7 +94,7 @@ public void run() {
 		return new Thread(exactlyOnceValidationConsumer);
 	}
 
-	private static class ExactlyOnceValidatingMapper implements FlatMapFunction<String, String>, Checkpointed<BitSet> {
+	private static class ExactlyOnceValidatingMapper implements FlatMapFunction<String, String> {
```
Comment (StefanRRichter) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/testutils/ExactlyOnceValidatingConsumerThread.java]: Ok, I will replace the logic by `ListCheckpointed`.

File: flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java [hunk 3]
```diff
@@ -122,6 +124,22 @@ protected RocksDBStateBackend getStateBackend() throws IOException {
 		return backend;
 	}
 
+	// small safety net for instance cleanups, so that no native objects are left
```
Comment (StefanRRichter) [inline flink-contrib/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackendTest.java]: This change actually belongs to the previous commit, so I will put it back there.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java (hunk 1 above)
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateUtil.java]: Not strictly, but I would like to keep it.

**Comment by StefanRRichter** [comment]
I have pushed an updated version with the changes that @zentol suggested and the additional cleanups suggested by @dawidwys.

@dawidwys I removed the code that you mentioned, but it is still possible that there more dead code in CEP (or even other places).

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerMigrationTest.java [hunk 4]
```diff
@@ -1,149 +0,0 @@
-/*
```
Comment (aljoscha) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerMigrationTest.java]: We can't remove this because there is another PR that updates this for Flink 1.2 and 1.3: #4565

Comment (StefanRRichter) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerMigrationTest.java]: I discussed with @aljoscha that this class can be resurrected when merging #4565

**Comment by StefanRRichter** [comment]
Thanks for the review @aljoscha ! Will merge now.
